From 0903594cb289f6438203e8b352558a5854056a12 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 26 Jun 2014 15:53:20 -0400
Subject: [fs] nfs: Fix memroy leak for double mounts

Message-id: <1403798077-28799-2-git-send-email-steved@redhat.com>
Patchwork-id: 84874
O-Subject: [RHEL7.1 PATCH 01/78] NFS: Fix memroy leak for double mounts
Bugzilla: 1113248
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Kinglong Mee <kinglongmee@gmail.com>

When double mounting same nfs filesystem, the devname saved in d_fsdata
will be lost.The second mount should not change the devname that
be saved in d_fsdata.

Signed-off-by: Kinglong Mee <kinglongmee@gmail.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upsteam Id:4dfc7fdb9e261d77818616782ae1841a80e0fc83
Bugzilla: 1113248
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/getroot.c b/fs/nfs/getroot.c
index 66984a9..b94f804 100644
--- a/fs/nfs/getroot.c
+++ b/fs/nfs/getroot.c
@@ -120,7 +120,8 @@ struct dentry *nfs_get_root(struct super_block *sb, struct nfs_fh *mntfh,
 
  security_d_instantiate(ret, inode);
  spin_lock(&ret->d_lock);
- if (IS_ROOT(ret) && !(ret->d_flags & DCACHE_NFSFS_RENAMED)) {
+ if (IS_ROOT(ret) && !ret->d_fsdata &&
+     !(ret->d_flags & DCACHE_NFSFS_RENAMED)) {
   ret->d_fsdata = name;
   name = NULL;
  }
-- 
1.7.1